micro-rdk/src/common/ota.rs

```diff
@@ -352,53 +363,97 @@ impl<S: OtaMetadataStorage> OtaService<S> {
                             {
-                                Ok(pair) => break pair,
+                                Ok(pair) => {
+                                    sender = Some(pair.0);
+                                    inner_conn = Some(pair.1);
+                                    retry = false;
+                                }
                                 Err(e) => {
                                     log::error!("failed to build http request: {}", e);
+                                    retry = true;
                                 }
                             }
                         }
                         Err(e) => {
                             log::error!("failed to create tcp stream: {}", e);
+                            retry = true;
                         }
                     }
                 }
                 Err(e) => {
                     log::error!("failed to create http connection: {}", e);
+                    retry = true;
                 }
             };
 
-            log::debug!(
-                "retry establishing http connection to {} in {} seconds",
-                &uri,
-                CONN_RETRY_SECS
-            );
-            Timer::after(Duration::from_secs(CONN_RETRY_SECS)).await;
-        };
-
-        let conn = Box::new(self.exec.spawn(async move {
-            if let Err(err) = conn.await {
-                log::error!("connection failed: {:?}", err);
+            if sender.is_none() || inner_conn.is_none() {
+                return Err(OtaError::Other(
+                    "invalid state, error path not handled".to_string(),
+                ));
+            }
+            let mut sender = sender.unwrap();
+            let inner_conn = inner_conn.unwrap();
+
+            if retry {
+                log::debug!(
+                    "retry establishing http connection to {} in {} seconds",
+                    &uri,
+                    CONN_RETRY_SECS
+                );
+                Timer::after(Duration::from_secs(CONN_RETRY_SECS)).await;
+                continue;
             }
-        }));
-
-        log::info!("ota connected, beginning download");
-        let request = Request::builder()
-            .method("GET")
-            .uri(uri)
-            .body(Empty::<Bytes>::new())
-            .map_err(|e| OtaError::Other(e.to_string()))?;
-        let response = sender
-            .send_request(request)
-            .await
-            .map_err(|e| OtaError::Other(e.to_string()))?;
 
-        // TODO(RSDK-9346): handle other success codes
-        if response.status() != 200 {
-            return Err(OtaError::Other(
-                format!("Bad Request - Status:{}", response.status()).to_string(),
-            ));
+            conn = Some(Box::new(self.exec.spawn(async move {
+                if let Err(err) = inner_conn.await {
+                    log::error!("connection failed: {:?}", err);
+                }
+            })));
```

Comment:
[question] I've read this block a few times now, I'm still not sure what it is doing. I know it isn't new, but I'd still like to understand it. Maybe the best action to take here is a comment for this block.

Reply:
`conn` is the underlying connection and acts like an engine that keeps IO moving, https://docs.rs/hyper/latest/hyper/client/conn/http2/struct.Connection.html.

This is why we maintain a handle to it and drop it after we've finished sending/receiving all the bytes of a given request/response

Reply:
added inline comment